scrape/scrape.go

```diff
@@ -1273,8 +1280,30 @@ loop:
 		if ok {
 			err = app.AddFast(ce.ref, t, v)
 			_, err = sl.checkAddError(ce, met, tp, err, &sampleLimitErr, &appErrs)
+			switch err {
+			case nil:
+				if tp == nil {
+					sl.cache.trackStaleness(ce.hash, ce.lset)
+				}
+				// todo: This smells funny.
+				if hasExemplar := p.Exemplar(&e); hasExemplar {
+					if err := exemplarApp.AddExemplar(ce.lset, t, e); err != nil {
+						if err == storage.ErrDuplicateExemplar {
+							level.Debug(sl.l).Log("msg", "Duplicate exemplar", "seriesLabels", ce.lset, "exemplar", e)
+						} else {
+							level.Debug(sl.l).Log("msg", "unexpected error", "error", err, "seriesLabels", ce.lset, "exemplar", e)
+						}
+						// we don't want to write the exemplar to the WAL if we're not going to store it
```

Comment:
This seems weird.

Reply:
The comment itself, or the way we're getting around adding the exemplar to the WAL if the exemplar storage is not going to keep the exemplar? 

If the latter, do you have any suggestions? I think we'd have to add the exemplar storage to the fanout storage, and have it implement no ops for the rest of the regular Appender interface.

Reply:
The comment and the continue seem weird, the logic feels wrong.

Reply:
@brian-brazil Just looking at this again. I think the `don't write the exemplar to the WAL if we're not going to add it to the exemplar storage` is valid, trying to think of a nicer way to do that though. I wrote this PR assuming we'd want to keep sample storage and exemplar storage decoupled, but I'm struggling to think of a way to do this nicely without coupling them more closely. 

The logic would be explicitly backwards from the current fanout storage functions, tsdb is usually primary and if primary storage fails we error out before making the calls to secondary storage, but in this case the exemplar storage is primary. Is a one off in fanout where tsdb is not called first acceptable?

Reply:
Why not layer it on top of the current storage, so the exemplar storage gets first short at exemplars but passes all the other calls down transparently?